### PR TITLE
replace `#[allow_internal_unstable]` with `#[rustc_allow_const_fn_unstable]` for `const fn`s

### LINKED ISSUE #69399: #[allow_internal_unstable] does not allow unstable library features when applied on functions
The `#[allow_internal_unstable]` attribute can only enable compiler features. Library features cannot be enabled. 

Examples:

```rust
#![feature(allow_internal_unstable)]
// #![feature(wrapping_next_power_of_two, const_transmute)]

#[allow_internal_unstable(wrapping_next_power_of_two)]
fn g(a: u64) -> u64 {
    a.wrapping_next_power_of_two()
}

#[allow_internal_unstable(const_transmute)]
const fn h(a: u64) -> i64 {
    unsafe { std::mem::transmute(a) }
}
```

* Expected: can be built without errors.
* Actual: still complains about missing `#![feature]`s.

Rustc version = `1.43.0-nightly (2020-02-22 436494b8f8008b600d64)`

### PR DESCRIPTION
**Pull request by liketechnik**
`#[allow_internal_unstable]` is currently used to side-step feature gate and stability checks.
While it was originally only meant to be used only on macros, its use was expanded to `const fn`s.

This pr adds stricter checks for the usage of `#[allow_internal_unstable]` (only on macros) and introduces the `#[rustc_allow_const_fn_unstable]` attribute for usage on `const fn`s.

This pr does not change any of the functionality associated with the use of `#[allow_internal_unstable]` on macros or the usage of `#[rustc_allow_const_fn_unstable]` (instead of `#[allow_internal_unstable]`) on `const fn`s (see https://github.com/rust-lang/rust/issues/69399#issuecomment-712911540).

Note: The check for `#[rustc_allow_const_fn_unstable]` currently only validates that the attribute is used on a function, because I don't know how I would check if the function is a `const fn` at the place of the check. I therefore openend this as a 'draft pull request'.

Closes rust-lang/rust#69399

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_expand/src/base.rs [hunk 1]
```diff
@@ -768,8 +768,13 @@ impl SyntaxExtension {
         name: Symbol,
         attrs: &[ast::Attribute],
     ) -> SyntaxExtension {
-        let allow_internal_unstable = attr::allow_internal_unstable(sess, &attrs)
-            .map(|features| features.collect::<Vec<Symbol>>().into());
+        let allow_internal_unstable = {
+            let mut feat_list = Vec::new();
+            attr::allow_internal_unstable(sess, &attrs).map(|features| feat_list.extend(features));
+            attr::rustc_allow_const_fn_unstable(sess, &attrs)
```
Comment (oli-obk) [inline compiler/rustc_expand/src/base.rs:774]: I don't think we need this here. we just want `allow_internal_unstable` here, it's about macro expansion after all.

File: compiler/rustc_mir/src/transform/check_consts/mod.rs [hunk 2]
```diff
@@ -79,7 +79,7 @@ pub fn is_lang_panic_fn(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
 
 pub fn allow_internal_unstable(tcx: TyCtxt<'tcx>, def_id: DefId, feature_gate: Symbol) -> bool {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/check_consts/mod.rs:80]: please also rename this function

File: compiler/rustc_passes/src/check_attr.rs [hunk 3]
```diff
@@ -791,6 +793,26 @@ impl CheckAttrVisitor<'tcx> {
...
+    /// Outputs an error for `#[allow_internal_unstable]` which can only be applied to macros.
+    /// (Allows proc_macro functions)
+    fn check_rustc_allow_const_fn_unstable(
+        &self,
+        attr: &Attribute,
+        span: &Span,
+        target: Target,
+    ) -> bool {
+        if let Target::Fn | Target::Method(_) = target {
+            // FIXME Check that this isn't just a function, but a const fn
```
Comment (oli-obk) [inline compiler/rustc_passes/src/check_attr.rs:806]: you an use `self.tcx.is_const_fn_raw(self.tcx.hir().local_def_id(hir_id))` to check whether the function should be treated as a const fn.

**Comment by oli-obk** [comment]
This is really clean, thanks for taking the care to build clear individual commits

**Comment by liketechnik** [comment]
Note: If I understood the contributing guide correctly, it is discouraged to change already pushed commits after review (e. g. interactive rebase). Instead it is preferred to add new commits containing the necessary changes, which I did now. If I got that wrong, I'm happy to rebase instead.

**Comment by oli-obk** [comment]
The rule is there to make reviewing easier (since in large PRs it can quickly become infeasible to review the entire PR again), but once everything is approved fixing up history is definitely ok.

So... I'd r+ this as it is, though if you want to fix up history, that's totally alright with me, but not neccessary. Please let me know what you prefer

**Comment by liketechnik** [comment]
I'd leave it as is this time, because the fix-up commits should be descriptive enough on their own.

**Pull request merged**